### PR TITLE
Add credentials support in CF UPS.

### PR DESCRIPTION
**Pull request by jerome-benoit**
PR needs to be reviewed. 
PR have a dependency on deployment configuration changes: https://github.com/LucasBrazi06/ev-scp-sap-it/pull/44
PR targeted first for QA **only** deployment.

Update: 
+ SCP landscapes are properly setup and configuration are updated;
+ Rename method to get CF app env; 
+ Rename UPS template file to store credentials; 
+ Add check for configuration sections existence before modifying them.

Signed-off-by: Jérôme Benoit <jerome.benoit@sap.com>

### CODE REVIEW

File: src/utils/Configuration.ts [hunk 1]
```diff
@@ -356,17 +392,38 @@ export default class Configuration {
...
-      this.config = JSON.parse(fs.readFileSync(`${global.appRoot}/assets/config.json`, 'utf8')) as ConfigurationData;
+    if (!Configuration.config) {
+      Configuration.config = JSON.parse(fs.readFileSync(`${global.appRoot}/assets/config.json`, 'utf8')) as ConfigurationData;
+    }
+    return Configuration.config;
+  }
+
+  private static getAppEnv(): AppEnv {
+    if (!Configuration.appEnv) {
+      Configuration.appEnv = getAppEnv();
```
Comment (LucasBrazi06) [inline src/utils/Configuration.ts:414]: Weird, same method name here 'getAppEnv'.
```
private static getAppEnv(): AppEnv {
    if (!Configuration.appEnv) {
      Configuration.appEnv = getAppEnv();
```
Have you tested the call?

**Comment by LucasBrazi06** [comment]
> **Do not merge until SCP is properly setup and configuration are updated.**

@jerome-benoit What should I do?
This PR is in the pipeline 'To be reviewed' which means I can merge it but at the same time there is no comments above that says 'You can merge, SCP has been setup properly'.
If this PR is not ready to be merged, don't move it in this pipeline.
Let me know.

File: src/utils/Configuration.ts (hunk 1 above)
Comment (jerome-benoit) [reply src/utils/Configuration.ts:414]: Yes, there's no name collision and the deployment test on the dev landscape are working fine. Only the push notification have not been tested. So I think that PR should only be merged:
+ when you plan to deploy only on the QA for more tests coverage;
+ when the UPS setup is done on all landscape;
+ when the SCP services dependency is updated on all manifest files for dev and QA landscape. And then prod;

**Comment by jerome-benoit** [comment]
> > **Do not merge until SCP is properly setup and configuration are updated.**
> 
> @jerome-benoit What should I do?
> This PR is in the pipeline 'To be reviewed' which means I can merge it but at the same time there is no comments above that says 'You can merge, SCP has been setup properly'.
> If this PR is not ready to be merged, don't move it in this pipeline.
> Let me know.

I do not know how to express that: 
+ I want a code review;
+ I do not want the PR to be merged, it's a critical change that require QA deployment first only.

Dev cross-review pipeline?

**Comment by LucasBrazi06** [comment]
> I do not know how to express that:
> 
> * I want a final code review;
> * I do not want the PR to be merged, it's a critical change that require QA deployment first only.
> 
> Dev cross-review pipeline?

Maybe just a comment to request me to review but not merge? :-)

**Review by LucasBrazi06 — CHANGES_REQUESTED** [review]
What bothers me here is that we have one strategy on one hand and another on the other:
- Here we use User Provided Service to store sensitive data
- For the Crypto key we store it in the DB

We have to choose a strategy and stick to it.

I would choose the DB storage because it allows much more flexibility when the Admin wants to change it. He can change the encoding key of the JWT token, the SMTP server, Log level...

And all the config.json data will be dumped in the Technical Settings updatable by Admin.

What do you think?

**Comment by jerome-benoit** [comment]
> What bothers me here is that we have one strategy on one hand and another on the other:
> 
> * Here we use User Provided Service to store sensitive data
> * For the Crypto key we store it in the DB
> 
> We have to choose a strategy and stick to it.
> 
> I would choose the DB storage because it allows much more flexibility when the Admin wants to change it. He can change the encoding key of the JWT token, the SMTP server, Log level...
> 
> And all the config.json data will be dumped in the Technical Settings updatable by Admin.
> 
> What do you think?

It's a requirement a store credentials outside configuration files and it ahs to be done before the end of January. A meeting have already been scheduled by Christopher with SAP SGS to check if it's done. 
First, we fulfil SAP SGS requirements to avoid issue with SAP and the standardisation process.
Then we move to per tenant settings with credentials ciphered without any urgence because we already store credentials in a SAP security standards compliant manner.

**Comment by LucasBrazi06** [comment]
> It's a requirement a store credentials outside configuration files and it has to be done before the end of January. A meeting have already been scheduled by Christopher with SAP SGS to check if it's done.
> First, we fulfil SAP SGS requirements to avoid issue with SAP and the standardisation process.
> Then we move to per tenant settings with credentials ciphered without any urgence because we already store credentials in a SAP security standards compliant manner.

Ok, let's proceed in 2 steps then.

File: src/assets/ups-emobility-credentials.json [hunk 2]
```diff
@@ -0,0 +1,12 @@
+{
```
[the remark this answers is not in the document]
Comment (LucasBrazi06) [reply src/assets/ups-emobility-credentials.json:1]: @jerome-benoit ups-emobility-credentials-template.json ?

**Pull request merged**